perf-tool/src/monitor.cpp

```diff
@@ -160,5 +160,39 @@ JNIEXPORT void JNICALL MonitorContendedEntered(jvmtiEnv *jvmtiEnv, JNIEnv *env,
             j["stackTrace"] = jMethods;
         }
     }
+    /* Get the thread name */
+    jvmtiThreadInfo threadInfo;
+    error = jvmtiEnv->GetThreadInfo(thread, &threadInfo);
+    if (check_jvmti_error(jvmtiEnv, error, "Unable to retrieve thread info."))
+    {
+        j["threadName"] = threadInfo.name;
+        error = jvmtiEnv->Deallocate((unsigned char*)(threadInfo.name));
+        check_jvmti_error(jvmtiEnv, error, "Unable to deallocate thread name.\n");
+        // Local JNI refs to threadInfo.thread_group and threadInfo.context_class_loader will be freed upon return
```

Comment:
Out of curiosity, how does this get freed?

Reply:
This comes from the JVMTI documentation:

"All references returned by JVM TI functions are local references--these local references are created during the JVM TI call. Local references are a resource that must be managed (see the JNI Documentation). When threads return from native code all local references are freed. Note that some threads, including typical agent threads, will never return from native code. A thread is ensured the ability to create sixteen local references without the need for any explicit management. For threads executing a limited number of JVM TI calls before returning from native code (for example, threads processing events), it may be determined that no explicit management is needed. However, long running agent threads will need explicit local reference management--usually with the JNI functions PushLocalFrame and PopLocalFrame. Conversely, to preserve references beyond the return from native code, they must be converted to global references. These rules do not apply to jmethodID and jfieldID as they are not jobjects. "